[IA-4459] Clean up runtimes stuck in deleting status for over an hour

LizBaldo opened the pull request:
Jira ticket: https://broadworkbench.atlassian.net/browse/IA-4459

Note that I still need to test, but I need to figure out how to run this on my BEE first :)

jdcanas commented on .github/workflows/trivy.yml:
```diff
@@ -35,6 +35,7 @@ jobs:
 
       # build the image
       - name: Build
+        continue-on-error: true
```
Hmm.. what does this accomplish?

LizBaldo replied:
This is just me testing the GHA, please ignore I will remove it soon

jdcanas commented on resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:
```diff
@@ -60,6 +61,23 @@ object DbReader {
           )"""
       .query[Runtime]
 
+  val deletingRuntimeQuery =
```
Seems like a different name is warranted since this gets runtimes in `Stopped` and `Running` as well. Maybe name it or leave a comment to indicate the usage as opposed to the statuses it includes.

LizBaldo replied:
I am not sure I understand, the goal is to grab runtimes marked as `deleting` in leo db, but would show as Running or Stopped in Google. Similar to the deletedRuntimeQuery and the erroredRuntimeQuery

Qi77Qi replied:
AoU re-uses runtime names...so just need to make sure we won't accidentally pick up runtimes that isn't in `Deleting`

Qi77Qi commented on .scalafmt.conf:
```diff
@@ -1,4 +1,4 @@
-version = 3.6.1
+version = 3.7.12
```
thanks for keeping things up-to-date!

Qi77Qi commented on resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DeletingRuntimeChecker.scala:
```diff
@@ -0,0 +1,104 @@
+import org.typelevel.log4cats.Logger
+
+// Implements CheckRunner[F[_], A]
+object DeletingRuntimeChecker {
```
maybe add a comment explaining what this checker intends to do

Qi77Qi commented on resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:
```diff
@@ -60,6 +61,23 @@ object DbReader {
+          INNER JOIN RUNTIME_CONFIG AS rt ON c1.runtimeConfigId = rt.id
+          WHERE
+            c1.status = "Deleting" AND
+            c1.destroyedDate > now() - INTERVAL 60 MIN AND
```
should this be `c1.dataAccessed < now() - INTERVAL 60 MIN`...basically making sure the runtime hasn't been accessed for more than 1 hr.

`destroyedDate` will always be the dummy tmestamp becuz these runtimes aren't deleted yet

LizBaldo replied:
Good point, I will do that, thanks!

LizBaldo commented on janitor/cloudbuild.yaml:
```diff
@@ -2,7 +2,7 @@ steps:
     args:
       - |
-        testOnly -- -l cronJobs.dbTest; janitor/docker:publishLocal
+        JVM_OPTS="-Xmx4G -Xms4G -Xss2M"; testOnly -- -l cronJobs.dbTest; janitor/docker:publishLocal
```
@Qi77Qi would you happen to know how I could pass these arguments to the docker container?

Qi77Qi replied:
try sth like 
```
  env:
  - JAVA_OPTS: -Xmx4G -Xms4G -Xss2M
 ```

see [this](https://cloud.google.com/build/docs/configuring-builds/substitute-variable-values) for setting env in cloud build

LizBaldo replied:
Ah thanks, looks slightly different from what I tried [here](https://github.com/broadinstitute/leonardo-cron-jobs/pull/289/commits/35e5e6db3c00d391280dfbad1a0f4bc3977f0eec), I will give it another try now :)

LizBaldo replied:
That worked, thanks a lot!

Qi77Qi replied:
yea, it should be `JAVA_OPTS`

LizBaldo commented:
Fixed previous DBReader unit tests that are not run by CI but have to be manually done locally, and added a test for this specific use case, things work well with my Local leo DB now :)

rtitle commented on core/src/main/scala/com/broadinstitute/dsp/checkerDeps.scala:
```diff
@@ -58,13 +63,14 @@ sealed abstract class Runtime {
 object Runtime {
   final case class AzureVM(id: Long,
-                           cloudContext: CloudContext.Azure,
+                           azureCloudContext: AzureCloudContext, //                           cloudContext: CloudContext.Azure,
```
Remove comment?

LizBaldo replied:
Thanks!

rtitle commented on resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala:
```diff
@@ -60,6 +61,23 @@ object DbReader {
+          WHERE
+            c1.status = "Deleting" AND
+            c1.dateAccessed < now() - INTERVAL 1 HOUR AND
+            NOT EXISTS (
```
Trying to understand why we need the `AND NOT EXISTS …` clause here. It seems unnecessary, since there is a unique constraint on `(cloudContext, name, destroyedDate)`?

LizBaldo replied:
From Qi's comment above, it is necessary precaution as AOU reuses names

rtitle replied:
I think that would only be necessary for `Deleted` runtimes though, otherwise there is a DB constraint.

Qi77Qi replied:
since the runtime is `Deleting`...I don't think `destroyedDate` will be defined yet... and (cloudContext, name) is not guranteed to be unique

rtitle replied:
Yeah we only set `destroyedDate` when the status is `Deleted`… which means the constraint on `(cloudContext, name, destroyedDate)` only allows reusing names when at most 1 is non-deleted. (So you can't have a Deleting and Running runtime with the same name at the same time.)

I'm fine with leaving it as a sanity check… I was also wondering why the status check is not complete though. For example what if there were 2 runtimes:
```
name | status
-----
myRuntimeName | Deleted
myRuntimeName | Stopping
```

Would the `deletedRuntimeSql` mistakenly flag `myRuntimeName` and delete it in GCP?

Qi77Qi replied:
I think the `NOT EXISTS` exclude `Stopped` as well (both `Stopped` and `Running` are "active" statuses basically)..

but yea, it's tricky...think we did this only becuz AOU reuses names...I'll bring it up to see if we can use unique names going forward for AOU

Qi77Qi replied:
created https://precisionmedicineinitiative.atlassian.net/browse/RW-10776

rtitle commented on resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DeletingRuntimeChecker.scala:
```diff
@@ -0,0 +1,104 @@
+            checkAzureRuntime(x, isDryRun)
+        }
+
+      private def checkDataprocCluster(runtime: Runtime.Dataproc, isDryRun: Boolean)(implicit
```
These methods seem the same as `DeletedRuntimeChecker`.. wonder if it could be made more DRY?

LizBaldo replied:
I thought about it too, but the decoupling made is so easy to  develop and test that I ended up liking the framework and kept it, but if you feel strongly about it, I am happy to refactor

rtitle reviewed (commented):
This isn't necessarily new your PR, but something in the `resource-validator` SQL looks off to me.

1. `deletedRuntimeQuery`, `erroredRuntimeQuery`, `deletingRuntimeQuery` -- the action is the same in all these cases (we delete the runtime), so should they be collapsed into a single query + stream?

2. The `AND NOT EXISTS` is probably only necessary for the `Deleted` case -- otherwise the DB constraint prevents reusing names. And even the `deletedRuntimeQuery` SQL doesn't seem quite correct -- it only checks duplicates in Stopped and Running status, what if there is a runtime with the same name in `Starting`/`Stopping` while the RV happens top run?

Qi77Qi commented:
I'm fine with as is...hopefull we'll be able to clean this up a bit down the line

rtitle reviewed (approved):
I think my comments were a little beyond the scope of this PR, so thumbing this… would be great to clean up though.
